cpp/dolfinx/fem/CoordinateElement.cpp

```diff
@@ -72,97 +59,78 @@ ElementDofLayout CoordinateElement::create_dof_layout() const
-void CoordinateElement::pull_back_nonaffine(
-    xt::xtensor<double, 2>& X, const xt::xtensor<double, 2>& x,
-    const xt::xtensor<double, 2>& cell_geometry, double tol, int maxit) const
+void CoordinateElement::pull_back_nonaffine(mdspan2_t X, cmdspan2_t x,
+                                            cmdspan2_t cell_geometry,
+                                            double tol, int maxit) const
 {
   // Number of points
-  std::size_t num_points = x.shape(0);
+  std::size_t num_points = x.extent(0);
   if (num_points == 0)
     return;
 
   const std::size_t tdim = mesh::cell_dim(this->cell_shape());
-  const std::size_t gdim = x.shape(1);
-  const std::size_t num_xnodes = cell_geometry.shape(0);
-  assert(cell_geometry.shape(1) == gdim);
-  assert(X.shape(0) == num_points);
-  assert(X.shape(1) == tdim);
-
-  xt::xtensor<double, 2> dphi({tdim, num_xnodes});
-  xt::xtensor<double, 2> Xk({1, tdim});
+  const std::size_t gdim = x.extent(1);
+  const std::size_t num_xnodes = cell_geometry.extent(0);
+  assert(cell_geometry.extent(1) == gdim);
+  assert(X.extent(0) == num_points);
+  assert(X.extent(1) == tdim);
+
+  std::vector<double> dphi_b(tdim * num_xnodes);
+  mdspan2_t dphi(dphi_b.data(), tdim, num_xnodes);
+
+  std::vector<double> Xk_b(tdim);
+  mdspan2_t Xk(Xk_b.data(), 1, tdim);
+
   std::array<double, 3> xk = {0, 0, 0};
-  xt::xtensor<double, 1> dX = xt::empty<double>({tdim});
-  xt::xtensor<double, 2> J({gdim, tdim});
-  xt::xtensor<double, 2> K({tdim, gdim});
-  xt::xtensor<double, 4> basis(_element->tabulate_shape(1, 1));
+
+  std::vector<double> dX(tdim);
+
+  std::vector<double> J_b(gdim * tdim);
+  mdspan2_t J(J_b.data(), gdim, tdim);
+
+  std::vector<double> K_b(tdim * gdim);
+  mdspan2_t K(K_b.data(), tdim, gdim);
+
+  namespace stdex = std::experimental;
+  using mdspan4_t = stdex::mdspan<double, stdex::dextents<std::size_t, 4>>;
+
+  const std::array<std::size_t, 4> bsize = _element->tabulate_shape(1, 1);
+  std::vector<double> basis_b(
+      std::reduce(bsize.begin(), bsize.end(), 1, std::multiplies{}));
+  mdspan4_t basis(basis_b.data(), bsize);
+  std::vector<double> phi(basis.extent(2));
+
   for (std::size_t p = 0; p < num_points; ++p)
   {
-    std::fill(Xk.begin(), Xk.end(), 0.0);
+    std::fill(Xk_b.begin(), Xk_b.end(), 0.0);
     int k;
     for (k = 0; k < maxit; ++k)
     {
-      _element->tabulate(1, std::span(Xk.data(), Xk.size()), {1, tdim}, basis);
+      _element->tabulate(1, Xk_b, {1, tdim}, basis_b);
 
       // x = cell_geometry * phi
-      auto phi = xt::view(basis, 0, 0, xt::all(), 0);
       std::fill(xk.begin(), xk.end(), 0.0);
-      for (std::size_t i = 0; i < cell_geometry.shape(0); ++i)
-        for (std::size_t j = 0; j < cell_geometry.shape(1); ++j)
-          xk[j] += cell_geometry(i, j) * phi[i];
+      for (std::size_t i = 0; i < cell_geometry.extent(0); ++i)
+        for (std::size_t j = 0; j < cell_geometry.extent(1); ++j)
+          xk[j] += cell_geometry(i, j) * basis(0, 0, i, 0);
```

Comment:
Could use mdspan and math::dot for this operation, but i guess declaring 2 new spans is as verbose as this.